### PR TITLE
MVP

### PR DESCRIPTION
**Pull request by TylerHernandez**
This is our MVP

### CODE REVIEW

**Review by dag13 — APPROVED** [review]

File: src/main/java/com/google/sps/data/CommentDatabase.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.gson.*;
+import java.util.ArrayList;
+import java.util.List;
+
+public class CommentDatabase {
+
+  private static DatastoreService getDatastore() {
+    return DatastoreServiceFactory.getDatastoreService();
```
Comment (albert-g3) [inline src/main/java/com/google/sps/data/CommentDatabase.java:30]: This got addressed, at least in some file, this should be setting a static class field.

Comment (albert-g3) [reply src/main/java/com/google/sps/data/CommentDatabase.java:30]: @yjzhang111 please take a look

**Review by yjzhang111 — APPROVED** [review]

File: src/main/java/com/google/sps/data/CommentDatabase.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.gson.*;
+import java.util.ArrayList;
+import java.util.List;
+
+public class CommentDatabase {
```
Comment (albert-g3) [inline src/main/java/com/google/sps/data/CommentDatabase.java:27]: class name

File: src/main/java/com/google/sps/data/CommentDatabase.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+  private static DatastoreService getDatastore() {
+    return DatastoreServiceFactory.getDatastoreService();
+  }
+
+  public static Comment createComment(
```
Comment (albert-g3) [inline src/main/java/com/google/sps/data/CommentDatabase.java:33]: saveComment

File: src/main/java/com/google/sps/data/CommentDatabase.java (hunk 2 above)
Comment (mutisog) [inline src/main/java/com/google/sps/data/CommentDatabase.java:27]: In general, let's document public entities (classes, methods, interfaces, etc). In particular, this class is different enough that it warrants a few words on how it is used.

File: src/main/java/com/google/sps/data/CommentDatabase.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+      long rating, String drink, String content, String store, String email) {
```
Comment (mutisog) [inline src/main/java/com/google/sps/data/CommentDatabase.java:34]: Is the content field required i.e. must the user provide content before the UI accepts the comment? Depending on what you decide, consider addressing in a later pull request.

File: src/test/java/com/google/sps/data/CommentDatabaseTest.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  }
+
+  @Test
+  public void testGetCommentByID() {
+    Comment commentA =
+        CommentDatabase.createComment(RATING_A, DRINK_A, CONTENT_A, STORE_A, EMAIL_A);
+    Comment commentFromDB = CommentDatabase.getCommentByID(commentA.getId());
+    Assert.assertEquals(commentA.getRating(), commentFromDB.getRating());
+    Assert.assertEquals(commentA.getDrink(), commentFromDB.getDrink());
+    Assert.assertEquals(commentA.getContent(), commentFromDB.getContent());
```
Comment (mutisog) [inline src/test/java/com/google/sps/data/CommentDatabaseTest.java:62]: Think about cases where the method commentA.getRating() has a bug. How would this test behave?
Consider being more explicit and directly comparing commentFromDB.getRating() to RATING_A, as this will also catch bugs in commentA.getRating().
Same for the other comparisons.

File: src/main/java/com/google/sps/data/CommentDatabase.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    commentEntity.setProperty("drink", drink);
+    commentEntity.setProperty("content", content);
+    commentEntity.setProperty("store", store);
+    commentEntity.setProperty("email", email);
+    getDatastore().put(commentEntity);
+    long commentID = commentEntity.getKey().getId();
+    return new Comment(commentID, rating, drink, content);
+  }
+
+  public static Comment getCommentByID(long commentID) {
```
Comment (mutisog) [inline src/main/java/com/google/sps/data/CommentDatabase.java:47]: Is this method used anywhere? AFAICT, it's only referenced in a test

File: src/test/java/com/google/sps/data/CommentDatabaseTest.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    helper.setUp();
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Test
+  public void testGetCommentByID() {
```
Comment (mutisog) [inline src/test/java/com/google/sps/data/CommentDatabaseTest.java:56]: Doesn't need to be part of this pull request, but consider adding a test where getCommentById returns null (that counts as an edge case).

File: src/main/java/com/google/sps/servlets/CommentServlet.java [hunk 8]
```diff
@@ -29,65 +25,42 @@
...
-
-    Entity taskEntity = new Entity("Task");
-    taskEntity.setProperty("comment", comment);
-    taskEntity.setProperty("timestamp", timestamp);
+    String drink = getParameter(request, "drink", "");
+    drink = drink.toLowerCase();
+    long rating = Integer.parseInt(getParameter(request, "rating", ""));
+    String content = getParameter(request, "content", "");
+    String store = getParameter(request, "store", "");
+    String email = getParameter(request, "email", "");
```
Comment (mutisog) [inline src/main/java/com/google/sps/servlets/CommentServlet.java:55]: Can be addressed in a subsequent pull request, but should we allow user comments without a store and email?

File: src/main/java/com/google/sps/servlets/DrinkRatingServlet.java [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+// Servlet to process users rating a drink from a coffee shop
+@WebServlet("/drink-rating")
+public class DrinkRatingServlet extends HttpServlet {
+
+  private final double maxRating = 5.0;
+  private List<String> knownDrinks = new ArrayList<>();
+  public Map<String, Integer> votes = createMap();
```
Comment (mutisog) [inline src/main/java/com/google/sps/servlets/DrinkRatingServlet.java:37]: what kind of votes? Perhaps a better name?

File: src/main/java/com/google/sps/servlets/DrinkRatingServlet.java [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+
+    for (int i = 1; i <= votes.size(); i++) {
+      String str = Integer.toString(i);
+      totalRatings = i * votes.get(str);
+      totalVotes += votes.get(str);
+    }
+
+    double average = (double) (totalRatings / totalVotes);
+    average = Math.round(average * 10) / 10.0;
+    return average;
```
Comment (mutisog) [inline src/main/java/com/google/sps/servlets/DrinkRatingServlet.java:62]: I thought the plan was to create a method with a name that clarifies we're rounding to one decimal place. Assume this will be done in a subsequent pull request?

File: src/main/java/com/google/sps/servlets/DrinkRatingServlet.java [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+import java.io.IOException;
+import java.util.*;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+// Servlet to process users rating a drink from a coffee shop
+@WebServlet("/drink-rating")
+public class DrinkRatingServlet extends HttpServlet {
```
Comment (mutisog) [inline src/main/java/com/google/sps/servlets/DrinkRatingServlet.java:33]: Would like to see tests for this class in a subsequent pull request.

File: src/main/java/com/google/sps/servlets/DrinkRatingServlet.java [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+
+  private final double maxRating = 5.0;
+  private List<String> knownDrinks = new ArrayList<>();
+  public Map<String, Integer> votes = createMap();
+
+  private Map<String, Integer> createMap() {
+    Map<String, Integer> map = new LinkedHashMap<>();
```
Comment (mutisog) [inline src/main/java/com/google/sps/servlets/DrinkRatingServlet.java:40]: The use of a String in this Map still seems awkward -- notice the number of times toString is used in subsequent methods, and could become a source of bugs. I suggest revisiting this in a subsequent pull request.

Comment (mutisog) [inline src/main/java/com/google/sps/servlets/DrinkRatingServlet.java:40]: Curious, why the choice to use the LinkedHashMap?

File: src/main/java/com/google/sps/data/CommentDatabase.java [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    long rating = (long) commentEntity.getProperty("rating");
+    String drink = (String) commentEntity.getProperty("drink");
+    String content = (String) commentEntity.getProperty("content");
+
+    return new Comment(commentID, rating, drink, content);
+  }
+
+  public static List<Comment> getCommentByStore(String storeID) {
```
Comment (albert-g3) [inline src/main/java/com/google/sps/data/CommentDatabase.java:65]: getCommenstByStore since returns a list

File: src/main/java/com/google/sps/servlets/DrinkRatingServlet.java (hunk 10 above)
Comment (mutisog) [inline src/main/java/com/google/sps/servlets/DrinkRatingServlet.java:62]: Can simplify to "return Math.round(...."

**Review by mutisog — APPROVED** [review]
Left some comments, which can be addressed in a later pull request. Let's aim to get this merged before today's sync.

File: src/main/java/com/google/sps/data/CommentDatabase.java [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+      String drink = (String) entity.getProperty("drink");
+      String content = (String) entity.getProperty("content");
+
+      commentByStore.add(new Comment(id, rating, drink, content));
+    }
+
+    return commentByStore;
+  }
+
+  public static List<Comment> getCommentByEmail(String email) {
```
Comment (albert-g3) [inline src/main/java/com/google/sps/data/CommentDatabase.java:83]: same as above should be comment**s**

File: src/main/java/com/google/sps/servlets/CommentServlet.java [hunk 15]
```diff
@@ -29,65 +25,42 @@
...
-    for (Entity entity : results.asIterable()) {
-      long id = entity.getKey().getId();
-      long timestamp = (long) entity.getProperty("timestamp");
-      String comment = (String) entity.getProperty("comment");
-
-      Task task = new Task(id, timestamp, comment);
-      tasks.add(task);
+    List<Comment> comments;
+    if (getParameter(request, "store", "").isEmpty()) {
+      comments = CommentDatabase.getCommentByEmail(getParameter(request, "email", ""));
```
Comment (albert-g3) [inline src/main/java/com/google/sps/servlets/CommentServlet.java:37]: @yjzhang111 I thought you changed getParameter so there is no need to pass "" as third parametere

File: src/main/java/com/google/sps/servlets/DrinkRatingServlet.java [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+
+    /* Initializes a map with 5 keys corresponding to the ratings and gives each
+     *  of the keys a value of 0 indicating how many people have made that rating so far
+     */
+    for (int i = 1; i <= 5; i++) map.put(Integer.toString(i), 0);
+    return map;
+  }
+
+  private double getAverageRating(Map<String, Integer> votes) {
```
Comment (albert-g3) [inline src/main/java/com/google/sps/servlets/DrinkRatingServlet.java:49]: not clear what the mappin is, votes to what?

**Review by albert-g3 — CHANGES_REQUESTED** [review]
